pkg/controllers/machinemigration/machine_migration_controller_test.go

```diff
@@ -0,0 +1,693 @@
+				mapiMachine = mapiMachineBuilder.
+					WithAuthoritativeAPI(machinev1beta1.MachineAuthorityMachineAPI).
+					Build()
+				Eventually(k8sClient.Create(ctx, mapiMachine)).Should(Succeed())
+
+				By("Setting the MAPI machine status AuthoritativeAPI to MachineAPI")
+				Eventually(k.UpdateStatus(mapiMachine, func() {
+					mapiMachine.Status.AuthoritativeAPI = machinev1beta1.MachineAuthorityMachineAPI
+				})).Should(Succeed())
+
+				req = reconcile.Request{NamespacedName: client.ObjectKeyFromObject(mapiMachine)}
+			})
+
+			It("should do nothing and not requeue", func() {
+				By("Running one reconciliation")
+				result, err := reconciler.Reconcile(ctx, req)
+				Expect(err).NotTo(HaveOccurred(), "reconciler should not have errored")
+				Expect(result.Requeue).To(BeFalse())
+			})
+		})
+
+		Context("when status.AuthoritativeAPI is empty (first observation)", func() {
+			BeforeEach(func() {
+				By("Setting the MAPI machine spec AuthoritativeAPI to MachineAPI")
+				mapiMachine = mapiMachineBuilder.
+					WithAuthoritativeAPI(machinev1beta1.MachineAuthorityMachineAPI).
+					Build()
+				Eventually(k8sClient.Create(ctx, mapiMachine)).Should(Succeed())
+
+				By("Leaving the MAPI machine status AuthoritativeAPI empty")
+
+				req = reconcile.Request{NamespacedName: client.ObjectKeyFromObject(mapiMachine)}
+			})
+
+			It("should patch the status to match spec and requeue", func() {
+				By("Running one reconciliation")
+				_, err := reconciler.Reconcile(ctx, req)
+				Expect(err).NotTo(HaveOccurred(), "reconciler should not have errored")
+				updatedM := &machinev1beta1.Machine{}
+				Expect(k8sClient.Get(ctx, client.ObjectKeyFromObject(mapiMachine), updatedM)).To(Succeed())
+				Expect(updatedM.Status.AuthoritativeAPI).To(Equal(updatedM.Spec.AuthoritativeAPI))
+			})
+		})
+
+		Context("when the Synchronized condition is not True", func() {
+			BeforeEach(func() {
+				By("Setting the MAPI machine spec AuthoritativeAPI to ClusterAPI")
+				mapiMachine = mapiMachineBuilder.
+					WithAuthoritativeAPI(machinev1beta1.MachineAuthorityClusterAPI).
+					Build()
+				Eventually(k8sClient.Create(ctx, mapiMachine)).Should(Succeed())
+
+				By("Setting the MAPI machine status AuthoritativeAPI to MachineAPI")
+				Eventually(k.UpdateStatus(mapiMachine, func() {
+					updatedMAPIMachine := mapiMachineBuilder.
+						WithAuthoritativeAPIStatus(machinev1beta1.MachineAuthorityMachineAPI).
+						WithConditions([]machinev1beta1.Condition{{
+							Type:               consts.SynchronizedCondition,
+							LastTransitionTime: metav1.Now(),
+							Status:             corev1.ConditionFalse}}).
+						Build()
+					mapiMachine.Status.AuthoritativeAPI = updatedMAPIMachine.Status.AuthoritativeAPI
+					mapiMachine.Status.Conditions = updatedMAPIMachine.Status.Conditions
+				})).Should(Succeed())
+
+				req = reconcile.Request{NamespacedName: client.ObjectKeyFromObject(mapiMachine)}
+			})
+
+			It("should requeue for later processing", func() {
```

Comment:
Should we check the machine hasn't been written to? Resource version hasn't changed?

Reply:
In the migration controllers we don't have logic that changes the spec, only the metadata (annotations) and the status. So we would never update the ResourceVersion right? In any case

Reply:
Resource version updates any time a resource is written, doesn't matter whether it's spec/status/metadata

Reply:
I misremembered!
Added Expects to cover these where appropriate.